naca0012-example (#741)
lethe-cfd/lethe: doc/source/examples/incompressible-flow/2d-naca0012-low-reynolds/2d-naca0012-low-reynolds.rst

===============================================
2D Flow around NACA0012 at low Reynolds number
===============================================

Features
--------

- Solver: ``gls_navier_stokes_2d`` (with Q1-Q1)
- Transient problem
- Boundary Layer Mesh - Transfinite Mesh
- Spectral analysis - Fourier transform

Files used in this example
----------------------------

- Base case (:math:`Re = 1000`): ``/examples/incompressible-flow/2d-naca0012-low-reynolds/naca.prm``
- .geo file to generate the mesh: ``/examples/incompressible-flow/2d-naca0012-low-reynolds/c_type_mesh.geo``

Description of the case
-------------------------

> In this example, a two-dimensional flow around a NACA airfoil is studied. According to Wikipedia, the *NACA airfoils are airfoil shapes for aircraft wings developed by the National Advisory Committee for Aeronautics (NACA)*. The NACA is the ancestor of the well known `NASA <https://www.nasa.gov/about/index.html>`_. The airfoils, simply referred to as NACAs afterwards, can be described by a set of four digits for the simpler airfoils, allowing to generate a lot of different profiles. Since we deal with deal with a symmetrical NACA we will only consider the last two digits which describe the maximum thickness of the airfoil in percentage of the chord. The first two digits, describing the maximum camber and its position, are set to zero. The problem under study is presented in the figure below, which gives details about the boundary conditions chosen and gives an idea of the size of the box for the resolution.

Reviewer: ```suggestion
In this example, a two-dimensional flow around a NACA airfoil is studied. According to Wikipedia, the *NACA airfoils are airfoil shapes for aircraft wings developed by the National Advisory Committee for Aeronautics (NACA)*. The NACA is the ancestor of the well-known `NASA <https://www.nasa.gov/about/index.html>`_. The airfoils, simply referred to as NACAs afterwards, can be described by a set of four digits for the simpler airfoils, allowing the generation of a lot of different profiles. Since we deal with a symmetrical NACA we will only consider the last two digits, which describe the maximum thickness of the airfoil in percentage of the chord. The first two digits, describing the maximum camber and its position, are set to zero. The problem under study is presented in the figure below, which details the boundary conditions chosen and gives an idea of the box size for the resolution.
```

.. image:: image/explanations_BC.png

> These airfoils, though created almost 100 years ago, are still very popular in aerodynamic fields where they often serve as a *base* choice, which is then modified to fit the constructor's constraints. The literature on numerical simulations of NACAs with high Reynolds numbers (:math:`Re>10^6`) is abundant. However recent technological developments in the field of Micro Air Vehicles (MAVs) created a need for low Reynolds simulations (:math:`Re<10^4`) since these vehicles operate at Reynolds numbers much smaller than usual aircraft. In this example, a 2D simulation around a 0012NACA at :math:`Re=1000` will be performed in order to retrieve significant quantities such as drag, lift and pressure coefficients, respectively denoted by :math:`C_D`, :math:`C_L` and :math:`C_p`. Elements of order 1 will be used for pressure and velocity. A spectral analysis will then be performed to analyze the vortex shedding frequency. 

Reviewer: ```suggestion
These airfoils, though created almost 100 years ago, are still very popular in aerodynamic fields where they often serve as a *base* choice, which is then modified to fit the constructor's constraints. The literature on numerical simulations of NACAs with high Reynolds numbers (:math:`Re>10^6`) is abundant. However, recent technological developments in the field of Micro Air Vehicles (MAVs) created a need for low Reynolds simulations (:math:`Re<10^4`) since these vehicles operate at Reynolds numbers much smaller than usual aircraft. In this example, a 2D simulation around a 0012NACA at :math:`Re=1000` will be performed in order to retrieve significant quantities such as drag, lift, and pressure coefficients, respectively denoted by :math:`C_D`, :math:`C_L` and :math:`C_p`. Elements of order 1 will be used for pressure and velocity. A spectral analysis will then be performed to analyze the vortex shedding frequency. 
```

Parameter file
----------------

Simulation control
~~~~~~~~~~~~~~~~~~~~~

> The problem is transient. An adaptative time step is used to ensure that the CFL:math:`<10` condition is obtained. This boundary for the CFL number was chosen to get relevant results while having reasonable simulation times. The reader is refered to the :doc:`../2d-lid‐driven-cavity-flow/lid‐driven-cavity-flow` example for a better understanding of the time step adaptation to kep the CFL below a certain treshold. A second order backward differentiation (BDF2) scheme was chosen.

Reviewer: ```suggestion
The problem is transient. An adaptative time step is used to ensure that the CFL:math:`<10` condition is obtained. This boundary for the CFL number was chosen to get relevant results while having reasonable simulation times. The reader is referred to the :doc:`../2d-lid‐driven-cavity-flow/lid‐driven-cavity-flow` example for a better understanding of the time step adaptation to keep the CFL below a certain threshold. A second-order backward differentiation (BDF2) scheme was chosen.
```

.. code-block:: text

    subsection simulation control
      set method            = bdf2
      set output name       = naca-output
      set output path       = ./output/
      set output frequency  = 50
      set adapt             = true
      set max cfl           = 10
      set time end          = 40
      set time step         = 0.01
    end
    
Physical properties 
~~~~~~~~~~~~~~~~~~~~~

In this problem, the Reynolds number is defined as:

.. math::
        Re = \frac{u_{\infty}C}{\nu}

> where :math:`C` is the chord length of the airfoil, :math:`u_{\infty}` is the magnitude of the inlet velocity (free-stream velocity) and :math:`\nu` the kinematic viscosity. :math:`C` and :math:`u_{\infty}` are taken equal to 1. Then, in order to get :math:`Re = 1000` one must have :math:`\nu = 0.001`.

Reviewer: ```suggestion
where :math:`C` is the chord length of the airfoil, :math:`u_{\infty}` is the magnitude of the inlet velocity (free-stream velocity) and :math:`\nu` the kinematic viscosity. :math:`C` and :math:`u_{\infty}` are taken equal to 1. Then, in order to get :math:`Re = 1000`, one must have :math:`\nu = 0.001`.
```

.. code-block:: text
	
    subsection physical properties
      subsection fluid 0
        set kinematic viscosity = 0.001
      end
    end

Forces
~~~~~~~~~~~~~~~~~~

The forces are computed in order to obtain the drag and lift coefficients later on:

.. code-block:: text
	
    subsection forces
      set verbosity             = verbose
      set calculate force       = true
      set calculate torque      = false
      set force name            = force
      set output precision      = 10
      set calculation frequency = 1
      set output frequency      = 1
    end
    
Mesh 
~~~~~~~~~~~~~~~~~~

> A C-Type mesh was created around the NACA. It is one of the usual types of mesh chosen for airfoils because it allows for the curvature of the grid to match the leading edge of the airfoil. Also, it allows the user to place more cells in areas which require higher resolution, typically near the upper and lower surfaces of the airfoil and in the wake. These properties are also true for O-type meshes, the difference being that the exterior boundary is a full circle, instead of a semi-circle in entry. In order to obtain such a mesh, the Transfinite functions of ``gmsh`` were used. To get a good understanding of how it was done, the reader is advised to read through the ``cTypeMesh.geo`` file included in the example which is thoroughly commented. To generate a mesh with a different angle of attack, the only thing required is to change the ``angle`` parameter in the ``cTypeMesh.geo`` file.

Reviewer: ```suggestion
A C-Type mesh was created around the NACA. It is one of the usual types of mesh chosen for airfoils because it allows for the curvature of the grid to match the leading edge of the airfoil. Also, it allows the user to place more cells in areas which require higher resolution, typically near the upper and lower surfaces of the airfoil and in the wake. These properties are also true for O-type meshes, the difference being that the exterior boundary is a full circle, instead of a semi-circle in entry. In order to obtain such a mesh, the Transfinite functions of ``gmsh`` were used. To get a good understanding of how it was done, the reader is advised to read through the ``cTypeMesh.geo`` file included in the example, which is thoroughly commented on. To generate a mesh with a different angle of attack, the only thing required is to change the ``angle`` parameter in the ``cTypeMesh.geo`` file.
```

.. code-block:: text
	
    subsection mesh
      set type      = gmsh
      set file name = naca.msh
    end

Below is the whole mesh and a zoom on the airfoil, for an angle of attack :math:`\alpha = 15°`

.. image:: image/whole_mesh.png

.. image:: image/zoomed_mesh_15.png

Mesh adaptation 
~~~~~~~~~~~~~~~~~~

> Mesh adaptation is used to get a higher resolution in areas of interest, that is to say close to the airfoil, while keeping a coarse mesh far from the NACA. Since the mesh is big and the simulation lengthy in time, it was chosen not to refine too much. Also, since the area of interest of the mesh (close to the airfoil) is much smaller than the whole mesh, the coarsening fraction was set eight times bigger than the refinement fraction. The parameters were tuned as follows: 

Reviewer: ```suggestion
Mesh adaptation is used to get a higher resolution in areas of interest, that is to say, close to the airfoil, while keeping a coarse mesh far from the NACA. Since the mesh is big and the simulation lengthy in time, it was chosen not to refine too much. Also, since the area of interest of the mesh (close to the airfoil) is much smaller than the whole mesh, the coarsening fraction was set eight times bigger than the refinement fraction. The parameters were tuned as follows: 
```

.. code-block:: text
	
    subsection mesh adaptation
      set type                 = kelly
      set variable             = velocity
      set fraction type        = number
      set max number elements  = 700000
      set max refinement level = 2
      set min refinement level = 0
      set frequency            = 5
      set fraction refinement  = 0.02
      set fraction coarsening  = 0.16
    end
    

Boundary conditions
~~~~~~~~~~~~~~~~~~~

The boundary conditions are defined as presented above : 

.. code-block:: text

     subsection boundary conditions
       set number = 4
       subsection bc 0
         set type = noslip
       end
       subsection bc 1
         set type = function
         subsection u
           set Function expression = 1
         end
         subsection v
           set Function expression = 0
         end
         subsection w
           set Function expression = 0
         end
       end
       subsection bc 2
         set type = slip
       end
       subsection bc 3
       	set type = outlet
       	set beta = 1.3
       end
     end
	

> The boundary 0, corresponding to the NACA wall, is a ``noslip`` boundary condition which sets the velocity to zero on the boundary. The boundary 1 is the inlet where the velocity field was chosen to be horizontal and unitary, to ensure that :math:`Re = 1000` is correct. It is represented in green on the figure. The boundary 2, in black on the image, correspond to the upper and lower walls are endowed with a ``slip`` boundary condition and finally boundary 3 is of type ``outlet`` with a parameter :math:`\beta = 1.3`. The reader is referred to the `Parameters Guide <https://lethe-cfd.github.io/lethe/parameters/cfd/linear_solver_control.html>`_ for more information about the :math:`\beta` parameter.

Reviewer: ```suggestion
The boundary 0, corresponding to the NACA wall, is a ``noslip`` boundary condition that sets the velocity to zero on the boundary. Boundary 1 is the inlet where the velocity field was chosen to be horizontal and unitary to ensure that :math:`Re = 1000` is correct. It is represented in green on the figure. Boundary 2, in black on the image, corresponds to the upper and lower walls are endowed with a ``slip`` boundary condition, and finally, boundary 3 is of type ``outlet`` with a parameter :math:`\beta = 1.3`. The reader is referred to the `Parameters Guide <https://lethe-cfd.github.io/lethe/parameters/cfd/linear_solver_control.html>`_ for more information about the :math:`\beta` parameter.
```

Non-linear solver
~~~~~~~~~~~~~~~~~

> The ``inexact_newton`` non-linear solver is used with a high ``tolerance``, since convergence can be hard to obtain for high Reynolds number. The choice of this solver was to reduce the cost of the simulation since it reuses the Jacobian matrix between iterations.

Reviewer: ```suggestion
The ``inexact_newton`` non-linear solver is used with a high ``tolerance``, since convergence can be hard to obtain for high Reynolds numbers. The choice of this solver was to reduce the cost of the simulation since it reuses the Jacobian matrix between iterations.
```

	
Running the simulations
-----------------------

The simulation can be launched using the following command:

.. code-block:: text

	gls_navier_stokes_2d naca.prm

It can also run in parallel using:

.. code-block:: text

	mpirun -np X gls_navier_stokes_2d naca.prm

with X the number of processor used to run it.
	
However, it is highly recommended to launch the simulation on a supercomputer. To launch on a desktop machine, the ``time end`` can be set to ``3.0`` to see the beginning of the simulation. However, to get relevant results about the forces, it is better to simulate at least for 10 seconds so that a pseudo-steady regime settles.

Results and discussion
-----------------------

The following average pressure and velocity fields are obtained for an angle of attack :math:`\alpha` such that :math:`\alpha \in \{0,5,7,9,11,15\}` : 

.. image:: image/average_pressure.png

.. image:: image/average_velocity.png

> It is already noticeable that the higher the angle of attack, the greater the pressure gradient. Following this observation, the lift coefficient :math:`C_L` is expected to increase with the angle of attack, until stall is reached. The variation of the lift and drag coefficients are given below with a comparison to the work of Kouser et al. `[1] <https://doi.org/10.1177/17568293211055656>`_. Both coefficients are computed using the following formula : 

Reviewer: ```suggestion
It is already noticeable that the higher the angle of attack, the greater the pressure gradient. Following this observation, the lift coefficient :math:`C_L` is expected to increase with the angle of attack until stall condition is reached. The variation of the lift and drag coefficients are given below with a comparison to the work of Kouser et al. `[1] <https://doi.org/10.1177/17568293211055656>`_. Both coefficients are computed using the following formula : 
```

.. math::
        C_L = \frac{F_L}{0.5\rho_{\infty} u_{\infty}^2 S} \; \; \; \; \; C_D = \frac{F_D}{0.5\rho_{\infty} u_{\infty}^2 S}
        

> with :math:`F_L` and :math:`F_D` respectively the lift and drag forces. Those forces can be obtained in the ``force.00.dat`` and post processed using the ``post_processing.py`` python file included in the folder of this example. :math:`S` represents a reference area, here it is equal to the product of the chord length :math:`C` (equal to 1 in this example) multiplied by a unitary transversal length.

Reviewer: ```suggestion
with :math:`F_L` and :math:`F_D`, respectively, the lift and drag forces. Those forces can be obtained in the ``force.00.dat`` and post-processed using the ``post_processing.py`` python file included in the folder of this example. :math:`S` represents a reference area; here, it is equal to the product of the chord length :math:`C` (equal to 1 in this example) multiplied by a unitary transversal length.
```

        
.. image:: image/cl_cd_results_plot.png

> There is a slight variation of the results obtained compared to the literature results. This may be due to the discontinuity at the trailing edge, or a resolution not precise enough. Still, the evolution observed is relevant.

Reviewer: ```suggestion
There is a slight variation of the results obtained compared to the literature results. This may be due to the discontinuity at the trailing edge or a resolution not precise enough. Still, the evolution observed is relevant.
```

One can also see the low velocity zones on the upper part of the airfoil which corresponds to the recirculating zone: the ``noslip`` condition on the NACA imposes a zero velocity condition on the fluid. The following streamline representation helps to see the movements of the fluid inside the recirculating zone : 

.. image:: image/naca_streamline_angles.png

> It can be observed that zones of recirculation form on the airfoil. This is due to two phenomena : first the flow outside of the boundary layer tends to "pull" it in its direction and the ``noslip`` boundary condition slows the fluid, then a positive pressure gradient, commonly referred to as adverse pressure gradient, on the upper surface pushes the fluid backwards. Following this, the boundary layer separates and a recirculation zone is formed. Below is represented the mean pressure coefficient :math:`C_p` on the airfoil with comparison to the literature. It is computed using the following formula : 

Reviewer: ```suggestion
It can be observed that zones of recirculation form on the airfoil. This is due to two phenomena : first the flow outside of the boundary layer tends to "pull" it in its direction, and the ``noslip`` boundary condition slows the fluid, then a positive pressure gradient, commonly referred to as adverse pressure gradient, on the upper surface pushes the fluid backward. Following this, the boundary layer separates, and a recirculation zone is formed. Below is represented the mean pressure coefficient :math:`C_p` on the airfoil with a comparison to the literature. It is computed using the following formula : 
```

For angles of attack :math:`\alpha\geq 9°`, the vortices start to detach from the airfoil. It can be seen using the instantaneous velocity fields. The velocit fields for each angle of attack, at t = 40 seconds are shown below :  

.. image:: image/instantaneous_velocity.png

In order to retrieve the frequency of the vortex shedding, one can look at the fluctuations of :math:`C_L`, as presented below for the case where :math:`\alpha=15°` was considered: 

.. image:: image/plot_cl_time.png

The best mathematical tool available to make a spectral analysis is a Fourier transform, which is performed below, with literature results (Kouser et al. (2021) `[1] <https://doi.org/10.1177/17568293211055656>`_) for comparison:

.. image:: image/fft_cl_comparison.png

> The fundamental frequency is :math:`f_1 = 0.72` Hz which gives a shedding period :math:`T = 1.39` s that is coherent with the instantaneous velocity field above.

Reviewer: ```suggestion
The fundamental frequency is :math:`f_1 = 0.72` Hz, which gives a shedding period :math:`T = 1.39` s that is coherent with the instantaneous velocity field above.
```

Possibilities for extension
------------------------------

- **High-order elements** : In order to get more precise results on the forces and the coefficients, Q2-Q2 elements may be used. It can be modified by setting ``set velocity order = 2`` and ``set pressure order = 2`` in the ``FEM`` subsection of ``naca.prm`` .

> - **Going 3D** : the mesh can be extruded into the third dimension. Some modifications will be required in the boundary conditions and getting the correct boundaries id is not trivial. However with periodic boundary conditions set on the sides of the box, spanwise effects can be taken into account, which should yield much better results. 

Reviewer: ```suggestion
- **Going 3D** : the mesh can be extruded into the third dimension. Some modifications will be required in the boundary conditions, and getting the correct boundaries id is not trivial. However, with periodic boundary conditions set on the sides of the box, spanwise effects can be taken into account, which should yield much better results. 
```

Parameter file
----------------

Simulation control
~~~~~~~~~~~~~~~~~~~~~

The problem to solve is transient. An adaptative timestep is used to ensure that the :math:`CFL<10` condition is obtained. This boundary for the CFL number was chosen to get relevant results wile having reasonable simulation times. A second order backward difference scheme was chosen.

.. code-block:: text

    subsection simulation control
      set method            = bdf2
      set output name       = naca-output
      set output path       = ./output/
      set output frequency  = 50
      set adapt             = true
      set max cfl           = 10
      set time end          = 40
      set time step         = 0.01
    end
    
Physical properties 
~~~~~~~~~~~~~~~~~~~~~

In this problem, the Reynolds number is defined as :

.. math::

>         Re = \frac{uC}{\nu}

Reviewer 2: ```suggestion
        Re = \frac{u_{\infty}C}{\nu}
```

where :math:`C` is the chord length of the airfoil, :math:`u` is the entry velocity and :math:`\nu` the kinematic viscosity. :math:`C` and :math:`u` are taken equal to 1. Then, in order to get :math:`Re = 1000` one must have :math:`\nu = 0.001`.

.. code-block:: text
	
    subsection physical properties
      subsection fluid 0
        set kinematic viscosity = 0.001
        set density             = 1.0
      end
    end

Forces
~~~~~~~~~~~~~~~~~~

> The forces are computed in order to obtain the drag and lift coefficients later on :

Reviewer 3: ```suggestion
The forces are computed in order to obtain the drag and lift coefficients later on:
```

      set output precision      = 10
      set calculation frequency = 1
      set output frequency      = 1
    end
    
Mesh 
~~~~~~~~~~~~~~~~~~

A C-Type mesh was created around the NACA. It is the usual type of mesh chosen for airfoils because it allows for the curvature of the grid to match the leading edge of the airfoil. Also, it allows the user to place more cells on areas which require higher resolution, typicaly the upper and lower part of the airfoil. In order to obtain such a mesh, the Transfinite functions of ``gmsh`` were used. To get a good understanding of how it was done, the reader is advised to read through the ``cTypeMesh.geo`` file included in the example which is thoroughly commented. To generate a mesh with a different angle, the only thing required is to change the ``angle`` parameter in the ``cTypeMesh.geo`` file.

.. code-block:: text
	
    subsection mesh
      set type      = gmsh
      set file name = naca.msh
    end

Below is the whole mesh and a zoom on the airfoil, for an angle of attack :math:`\alpha = 15°`

.. image:: image/wholeMesh.png

.. image:: image/zoomedMesh15.png

Mesh adaptation 
~~~~~~~~~~~~~~~~~~

> Mesh adaptation is used to get a higher resolution in areas of interest, that is to say close to the airfoil, while keeping a coarse mesh far from the NACA. Since the mesh is big and the simulation lengthy in time, it was chosen not to refine too much. The parameters were tuned as follow : 

Reviewer: ```suggestion
Mesh adaptation is used to get a higher resolution in areas of interest, that is to say, close to the airfoil, while keeping a coarse mesh far from the NACA. Since the mesh is big and the simulation lengthy in time, it was chosen not to refine too much. The parameters were tuned as follows: 
```

.. code-block:: text
	
    subsection mesh adaptation
      set type                 = kelly
      set variable             = velocity
      set fraction type        = number
      set max number elements  = 700000
      set max refinement level = 2
      set min refinement level = 0
      set frequency            = 5
      set fraction refinement  = 0.02

>       set fraction coarsening  = 0.16

Reviewer 3: such a low refinement fraction vs coarsening fraction. I s there a reason for this? if so it is worth explaining in the text :)

    end
    
    
FEM
~~~

In this example, the interpolation order has been set to one for both velocity and pressure.

.. code-block:: text

    subsection FEM
      set pressure order = 1
      set velocity order = 1

>     end

Reviewer 3: These are the default value, is it worth it to redefine them?

Author: Should I remove this block code part and just say that the orders are 1 for velocity and pressure?

Reviewer 3: Yup, that's what I would suggest

Boundary conditions
~~~~~~~~~~~~~~~~~~~

The boundary conditions are defined as presented above : 

.. code-block:: text

     subsection boundary conditions
       set number = 4
       subsection bc 0
         set type = noslip
       end
       subsection bc 1
         set type = function
         subsection u
           set Function expression = 1
         end
         subsection v
           set Function expression = 0
         end
         subsection w
           set Function expression = 0
         end
       end
       subsection bc 2
         set type = slip
       end
       subsection bc 3
       	set type = outlet

>        	set beta = 1.3

Reviewer 3: 1.3 is a very odd value, why 1.3 and not like 10 or 100?

Author: Since it is an outlet condition, "beta should be close to unity" according to the Lethe doc on boundary conditions, so I thought 1.3 would be fitting and it proved to be working afterwards, so I decided not to change it.

Reviewer 3: Ok, it's a bit random but no biggy.

       end
     end
	

> The boundary 0 corresponds to the NACA, it is a ``noslip`` boundary condition which sets the velocity to zero on the boundary. The boundary 1 is the inlet where the velocity field was chosen to be horizontal and unitary, to ensure that :math:`Re = 1000` is correct, it is represented in green on the image. The boundary 2, in black on the image, is a wall with a ``slip`` boundary condition and finally boundary 3 is of type ``outlet`` with a parameter :math:`\beta = 1.3`. The reader is refered to the `Parameters Guide <https://lethe-cfd.github.io/lethe/parameters/cfd/linear_solver_control.html>`_ for more information about the :math:`\beta` parameter.

Author: Yes! I've added the precision

Non-linear solver
~~~~~~~~~~~~~~~~~

The ``inexact_newton`` non-linear solver is used with a high ``tolerance``, since convergence can be hard to obtain for high Reynolds number. The choice of this solver was to reduce the cost of the simulation since it reuses the Jacobian matrix between iterations.

.. code-block:: text

    subsection non-linear solver
      set solver         = inexact_newton
      set verbosity      = verbose
      set tolerance      = 1e-3
      set max iterations = 10
    end

Linear solver
~~~~~~~~~~~~~

Again, in order to reduce the computational time the tolerance for the linear solver were chosen higher than usual : 

.. code-block:: text

    subsection linear solver
      set verbosity                             = verbose
      set method                                = gmres
      set max iters                             = 5000
      set relative residual                     = 1e-3
      set minimum residual                      = 1e-8

>       set ilu preconditioner fill               = 1.1

Reviewer 3: Why use a fill of 1.1? It's a bit random, I don't think we have ever used non-integer fill level. That's weird :)

Author: It is a remnant of me playing with the parameters and should be ignored. Hopefully the results were not impacted by this, I'll suggest using "set ilu preconditioner fill               = 1"

      set ilu preconditioner absolute tolerance = 1e-10
      set ilu preconditioner relative tolerance = 1.00
      set max krylov vectors                    = 1000
    end      
	
	
.. tip::
	It is important to note that the ``minimum residual`` of the linear solver is smaller than the ``tolerance`` of the nonlinear solver. The reader can consult the `Parameters Guide <https://lethe-cfd.github.io/lethe/parameters/cfd/linear_solver_control.html>`_ for more information.
	
Running the simulations
-----------------------

The simulation can be launched using the following command 

.. code-block:: text

> 	gls_navier_stokes_2d nacaBatch.prm

Reviewer 3: Maybe specify the mpirun -np so user can launch it in parallel

	
However, it is highly recommended to launch the simulation on a supercomputer. To launch on a desktop machine, the ``time end`` can be set to ``3.0`` to see the beginning of the simulation. Though to get relevant results about forces, it is better to simulate at least for ``10s`` in order for a pseudo-steady regime to settle.

Results and discussion
-----------------------

The following average pressure and velocity fields are obtained for an angle of attack :math:`\alpha`such that :math:`\alpha \in \{0,5,7,9,11,15\}` : 

.. image:: image/pressionMoy.png

.. image:: image/vitesseMoy.png

It is already noticeable that the the bigger the angle of attack, the bigger the pressure gradient. Following this observation, the lift coefficient :math:`C_L` is expected to increase with the angle of attack. The variation of the lift and drag coefficients are given below with a comparison to the litterature (Kouser et al. (2021) `[1] <https://doi.org/10.1177/17568293211055656>`_), they are computed using the following formula : 

.. math::

>         C_L = \frac{F_L}{0.5\rho u^2 S} \; \; \; \; \; C_D = \frac{F_D}{0.5\rho u^2 S}

Reviewer 2: ```suggestion
        C_L = \frac{F_L}{0.5\rho u_{\infty}^2 S} \; \; \; \; \; C_D = \frac{F_D}{0.5\rho u_{\infty}^2 S}
```

        
with :math:`F_L` and :math:`F_D` respectively the lift and drag forces. Those forces can be obtained in the ``force.00.dat`` and post processed using the ``postProcessing.py`` python file included in the folder of this example. :math:`S` represents a reference area, here it is equal to the product of the chord length :math:`C` (equal to 1 in this example) multiplied by a unitary transversal length.
        
.. image:: image/ClCdResultsPlot.png

> There is a slight variation of the results obtained compared to the litterature results. Indeed, the simulation is purely two-dimensionnal while Kouser et al. take a 2D slice of a 3D simulation. Still, the evolution observed is relevant.

Reviewer 3: also what is a 2D slice when it comes to force? They calculated the force on a 2D slice or the mesh is a 2D extrusion into 3D?

> There is a slight variation of the results obtained compared to the litterature results. Indeed, the simulation is purely two-dimensionnal while Kouser et al. take a 2D slice of a 3D simulation. Still, the evolution observed is relevant.

Author: It is very true, that was a mistake from me, it is indeed a 2D simulation

> One can also see the low velocity zones on the upper part of the airfoil which corresponds to the boundary layer : the ``noslip`` condition on the NACA imposes a zero velocity condition on the fluid. This create a zone where the advective and diffusive forces are in competition. The following streamline representation helps to see the movements of the fluid inside the boundary layer : 

Reviewer: I would talk about the recirculating zone instead of the boundary layer.

.. image:: image/nacaStreamlineAngles.png

It can be observed that zones of recirculation form on the airfoil. This is due to two phenomenas : first the flow outside of the boundary layer tends to "pull" it in its direction and the ``noslip`` boundary condition slows the fluid, then a positive pressure gradient -commonly refered to as adverse pressure gradient- on the airfoil surface pushes the fluid backwards. Following this a vertex forms. Below is represented the mean pressure coefficient :math:`C_p` on the airfoil with comparison to the litterature right below. It is computed using the following formula : 

.. math::
        C_p = \frac{p-p_{\infty}}{0.5\rho_{\infty}u_{\infty}^2}
        
with :math:`p_{\infty}` the static pressure in the freestream (equal to 0 in this case), :math:`\rho_{\infty}` the freestream fluid density, equal to the fluid density since we are solving an incompressible flow and :math:`u_{\infty}` the freestream velocity of the fluid, equal to ``1.0`` in this case.

.. image:: image/CpPlot.png

> .. image:: image/cpPlotRef.png

Reviewer 2: You can use Webplot Digitizer to extract the data of the reference graph and compare the results in a single figure :)

Reviewer 2: Also, you can add the reference of the literature

Author: Reference added! 
 I thought it would be more readable to have two separate pressure coefficient plots since only the general for of it mattered to me

> The important pressure at the attacking edge of the airfoil is what allows the incoming flow to be deflected to the upper and lower part. Then, if we look at the upper part (be careful about the reversed y-axis since :math:`-C_p` is plotted) the adverse pressure gradient is visible. Then at the trailing edge, the mesh is ot precise enough. A slight discontuinity ensues which, though not physically accurate, do not invalidate the whole result.

Reviewer 2: ```suggestion
The important pressure at the leading edge of the airfoil is what allows the incoming flow to be deflected to the upper and lower part. Then, if we look at the upper surface (be careful about the reversed y-axis since :math:`-C_p` is plotted) the adverse pressure gradient is visible. Then at the trailing edge, the mesh is not precise enough. A slight discontinuity ensues which, though not physically accurate, do not invalidate the whole result.
```

For angles of attack :math:`\alpha\geq 9°`, the vortex start to detach from the airfoil, this can be seen using the instantaneous velocity fields. For each angle of attack, the last time iteration was taken :  

.. image:: image/vI.png

> In order to retrieve the frequency of the vortex shedding, one can look at the variations of :math:`C_L`. Below are the variations of the lift coefficient with time, only the case where :math:`\alpha=15°` was considered for this analysis : 

Reviewer 2: ```suggestion
In order to retrieve the frequency of the vortex shedding, one can look at the variations of :math:`C_L`. Below are the variations of the lift coefficient with time for the case where :math:`\alpha=15°` was considered for this analysis : 
```

Reviewer 3: variations could also be replaced by fluctuations

.. image:: image/plotClTemps.png

> The best mathematical tool available to make a spectral analysis is a Fourier transform, which is performed below, with literature results for comparison :

Reviewer 3: ```suggestion
The best mathematical tool available to make a spectral analysis is a Fourier transform, which is performed below, with literature results for comparison:
```
